packages/queued-request-controller/src/QueuedRequestMiddleware.ts

```diff
@@ -0,0 +1,140 @@
+    | NetworkControllerFindNetworkClientIdByChainIdAction
+    | SelectedNetworkControllerSetNetworkClientIdForDomainAction
+    | AddApprovalRequest,
+    never
+  >,
+  useRequestQueue: () => boolean,
+): JsonRpcMiddleware<any, any> => {
+  return createAsyncMiddleware(async (req: any, res: any, next) => {
+    if (!useRequestQueue() || isConfirmationMethod(req.method)) {
+      next();
+      return;
+    }
+
+    const networkClientIdForRequest = req.networkClientId as NetworkClientId;
+
+    await messenger.call(
+      QueuedRequestControllerActionTypes.enqueueRequest,
+      async () => {
+        if (req.method === 'wallet_switchEthereumChain') {
+          // eslint-disable-next-line n/callback-return
+          return next();
+        }
+
+        const isBuiltIn = isNetworkType(networkClientIdForRequest);
+        let networkConfigurationForRequest;
+        if (isBuiltIn) {
+          const builtIn = BUILT_IN_NETWORKS[networkClientIdForRequest];
+          if (builtIn.chainId) {
+            networkConfigurationForRequest = builtIn;
+          }
+        }
+
+        networkConfigurationForRequest =
+          networkConfigurationForRequest ||
+          messenger.call(
+            'NetworkController:getNetworkClientById',
+            networkClientIdForRequest,
+          ).configuration;
+
+        const currentProviderConfig = messenger.call(
+          'NetworkController:getState',
+        ).providerConfig;
+
+        const currentChainId = currentProviderConfig.chainId;
+
+        if (currentChainId === networkConfigurationForRequest.chainId) {
+          // eslint-disable-next-line n/callback-return
+          return next();
+        }
+
+        // if is switch eth chain call
+        // clear request queue when the switch ethereum chain call completes (success or fail)
+        // This is because a dapp-requested switch ethereum chain invalidates any requests they've made after this switch, since we dont know if they were expecting the chain after the switch or before.
+        // with the queue batching approach, this would mean clearing any batch for that origin (batches being per-origin.)
+        const requestData = {
+          toNetworkConfiguration: networkConfigurationForRequest,
+          fromNetworkConfiguration: currentProviderConfig,
+        };
+
+        try {
+          const approvedRequestData = await messenger.call(
+            'ApprovalController:addRequest',
+            {
+              origin,
+              type: ApprovalType.SwitchEthereumChain,
+              requestData,
+            },
+            true,
+          );
+
+          if (isBuiltIn) {
+            await messenger.call(
+              'NetworkController:setProviderType',
+              (approvedRequestData as { type: InfuraNetworkType }).type,
+            );
+          } else {
+            await messenger.call(
+              'NetworkController:setActiveNetwork',
+              (approvedRequestData as { id: NetworkClientId }).id,
```

Comment:
Here's one way we could get much better intellisense for these types. I'm not 100% sure the usage of the `ApprovalRequest` type is accurate here.

```suggestion
          )) as ApprovalRequest<typeof requestData> & {
            type: InfuraNetworkType;
          };

          if (isBuiltIn) {
            await messenger.call(
              'NetworkController:setProviderType',
              approvedRequestData.type,
            );
          } else {
            await messenger.call(
              'NetworkController:setActiveNetwork',
              approvedRequestData.id,
```

Reply:
See https://github.com/MetaMask/core/pull/1862